test/map.js

```diff
@@ -31,6 +31,23 @@ describe('map', function() {
     eq(h(10), (10 * 2) - 1);
   });
 
+  it('interprets (async (->) r) as a functor', function() {
+    var f = function(a) { return a.then(function(a) { return a - 1; }); };
+    var g = async function(b) { return b * 2; };
+    var h = R.map(f, g);
+    return h(10)
+      .then(h => {
+        eq(h, (10 * 2) - 1);
+      });
+  });
+
+  it('interprets generator as a functor', function() {
+    var f = function(gen) { return [gen.next().value - 1, gen.next().value - 2, gen.next().value - 3]; };
+    var g = function* (b) { yield b * 1; yield b * 2; yield b * 3; };
+    var h = R.map(f, g);
+    eq(h(10), [(10 * 1) - 1, (10 * 2) - 2, (10 * 3) - 3]);
+  });
```

Comment:
This test looks over-complicated to me.  Could we switch to two `yield`s/`next`s instead of three?

But it also makes me wonder whether a generator really is a functor.  Statefunesslscares me here.  But I'm a little foggy-brained (possible flu) to think it through right now.

Reply:
That's a good question... to my understanding, in terms of the way a generator behaves (ignoring any internal implementation details), a generator function is a pure function that returns an object that has state and an impure method (giving you different results each time you call .next()).
(a generator object: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/throw )

In the test example, you could invoke `g` as many times as you like and get a new generator - it's no different to `g` returning an object that has a function that may or may not have a side effect.
`map` itself doesn't interact with the generator object in any way, it just passes it through to `f`.

I can simplify the test to just two `yield`s, but if you think the impure nature is distracting from the proof that to map, 'GeneratorFunctions' are equivalent to 'Functions', then it can be replaced altogether with just one, something like this:
```
    var f = function(gen) { return gen.next().value - 1 };
    var g = function* (b) { yield b * 2; };
    var h = R.map(f, g);
    eq(h(10), (10 * 2) - 1);
```
which doesn't have any typically generator-y behaviour, but gives it much more symmetry with the original function-as-functor test "interprets (async (->) r) as a functor" above.

Reply:
I definitely prefer this alternate demonstration.

My brain is less foggy now (not much more than it's usual fogginess!)  And I still can't quite convince myself that generator functions obey the composition rule for functors.  But I'm also not coming up with a counterexample, so maybe I'm still more foggy than I'd like to think! :stuck_out_tongue_closed_eyes: 

Still, I this looks good to me.  Anyone have objections, @ramda/core?

Reply:
Thanks for the feedback - I have simplified the test with https://github.com/ramda/ramda/pull/2675/commits/f3cbb4be1843f3dd36dd1fcf119fba2af2b140ff